### PR TITLE
Add discord and bitbucket provider

### PR DESCRIPTION
**Pull request by afi-dev**
# Description

This pull request adds support for 🗝[Discord](https://discord.com) and 🗝[Bitbucket](https://bitbucket.org) single sign-on (SSO) providers to **fastapi-sso**. These additions extend authentication options. I added it because I personally needed it 😊, and I think it will be useful to others.

# Changes Made

- **Discord SSO Provider**: A new integration has been added to allow users to sign by using their **Discord accounts**. See [fastapi_sso/sso/discord.py](https://github.com/afi-dev/fastapi-sso/blob/master/fastapi_sso/sso/discord.py)

discordsso

- **Atlassian Bitbucket SSO Provider**: A new integration has been added to allow users to sign by using their **BitBucket accounts**. See [fastapi_sso/sso/bitbucket.py](https://github.com/afi-dev/fastapi-sso/blob/master/fastapi_sso/sso/bitbucket.py)

bitbucketsso

# Tests Performed

- [x] Pytest 👌
- [x] Personal integration test 👌

# Typechecking

- [x] black 👌
- [x] pylint 👌
- [x] mypy 👌
- [x] pyclean 👌
- [x] pre-commit 👌

# Exemples

- [x] MKDocs has been updated 👍

Examples of use have been added :
- Discord [exemples/discord.py](https://github.com/afi-dev/fastapi-sso/blob/master/examples/discord.py)
- BitBucket [exemples/bitbucket.py](https://github.com/afi-dev/fastapi-sso/blob/master/examples/discord.py)

### CODE REVIEW

File: docs/generate_reference.py [hunk 1]
```diff
@@ -1,39 +1,47 @@
 """Generate reference pages for the documentation."""
 
 from pathlib import Path
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, List, Dict, Union, Any, TypedDict
 
 if TYPE_CHECKING:
-    import mkdocs.config.defaults  # pragma: no cover
-
+    pass  # pragma: no cover
```
Comment (tomasvotava) [inline docs/generate_reference.py:7]: I believe this block can now be deleted :)

File: fastapi_sso/sso/discord.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    async def get_discovery_document(self) -> DiscoveryDocument:
+        return {
+            "authorization_endpoint": "https://discord.com/oauth2/authorize",
+            "token_endpoint": "https://discord.com/api/oauth2/token",
+            "userinfo_endpoint": "https://discord.com/api/users/@me",
+        }
+
+    async def openid_from_response(self, response: dict, session: Optional["httpx.AsyncClient"] = None) -> OpenID:
+        id = response.get("id")
```
Comment (tomasvotava) [inline fastapi_sso/sso/discord.py:43]: If possible, I'd avoid overriding builtins, maybe use `user_id` instead? I know there an 'id' attribute in the pydantic model, but in the properties scope it's not that visible, I'd say.

**Review by tomasvotava — CHANGES_REQUESTED** [review]
Thanks a lot, whant an awesome contribution! Could you please take a look at those two minor nitpicks I found? Also the pipeline is failing, could you take a look at that? I don't understand why the similarities check does not ignore function annotations, feel free to add ignore comment. If you need help with anything, let me know. Nice work!

File: docs/generate_reference.py (hunk 1 above)
Comment (afi-dev) [reply docs/generate_reference.py:7]: Yes indeed

Comment (afi-dev) [reply docs/generate_reference.py:7]: Indeed, I hadn't seen it, I remember that during the pylint I had an error in docs/generate_reference.py, I let copilot correct it without checking behind.

File: fastapi_sso/sso/discord.py (hunk 2 above)
Comment (afi-dev) [reply fastapi_sso/sso/discord.py:43]: Ok, yes we can manage that, I just put id because it seemed logical but it works too. I'm going to change that.

Comment (tomasvotava) [reply fastapi_sso/sso/discord.py:43]: It seems logical to me as well, I was just trying to find something 😁

**Comment by afi-dev** [comment]
I think I've applied the changes correctly, but please let me know if you need any further changes.

File: fastapi_sso/sso/discord.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            "token_endpoint": "https://discord.com/api/oauth2/token",
+            "userinfo_endpoint": "https://discord.com/api/users/@me",
+        }
+
+    async def openid_from_response(self, response: dict, session: Optional["httpx.AsyncClient"] = None) -> OpenID:
+        user_id = response.get("id")
+        avatar = response.get("avatar")
+        picture = None
+        if id and avatar:
+            picture = f"https://cdn.discordapp.com/avatars/{id}/{avatar}.png"
```
Comment (tomasvotava) [inline fastapi_sso/sso/discord.py:47]: This still needs to be replaced, I believe:

```suggestion
        if user_id and avatar:
            picture = f"https://cdn.discordapp.com/avatars/{user_id}/{avatar}.png"
```

**Review by tomasvotava — CHANGES_REQUESTED** [review]
Please take a look at the last thing, also, the pipeline is still failing, I am more of a ruff fan lately, so ignore it for now, I'll deal with it.

File: fastapi_sso/sso/discord.py (hunk 3 above)
Comment (afi-dev) [reply fastapi_sso/sso/discord.py:47]: Indeed, I went a little fast, I was a little preoccupied and I didn't modify everything very well.

**Comment by tomasvotava** [comment]
It looks perfect now, could you please try rebasing from the `master`? I switched to ruff, I hope it won't break anything, but the pipeline should now pass.

```
git checkout master
git pull
git checkout -
git rebase master
```

That should do it. There will probably be some conflicts, I can help you resolve them if you have any troubles.

**Comment by tomasvotava** [comment]
I felt bad for requesting so many changes, so I went ahead and created a PR into your PR https://github.com/afi-dev/fastapi-sso/pull/2 😁  it's all rebased, you only have to accept all incoming to resolve the conflicts. If you'd like to resolve yourself, feel free to close the pull request.

**Comment by afi-dev** [comment]
no problem

File: fastapi_sso/__init__.py [hunk 4]
```diff
@@ -18,6 +18,8 @@
 from .sso.notion import NotionSSO
 from .sso.spotify import SpotifySSO
 from .sso.twitter import TwitterSSO
+from .sso.bitbucket import BitbucketSSO
+from .sso.import DiscordSSO
```
[the remark this answers is not in the document]
Comment (afi-dev) [reply fastapi_sso/__init__.py]: 🥲 I'm sorry again, I'm a bit clumsy, these last few days I've been up all night working on my projects and I don't even know what I'm doing anymore, and I leave things like this.

**Review by afi-dev — COMMENTED** [review]
For me it's fine, I hope it's the last fix and that I haven't forgotten anything, after checking, there's nothing missing.

**Comment by tomasvotava** [comment]
Hey @afi-dev, this PR turned out to be such a failure on my side, I am so, so much sorry. If this is still relevant, is it ok if I take your commits so that you can keep the contribution, rebase it with latest changes and remove the changes to the docs generation as it's not relevant to Discord and Bitbucket?

**Pull request closed without merging**